### ISSUE TITLE
poor spacing in fraction inside square root (depends on rendering environment)

### ISSUE THREAD

**Issue by jason-s** [KaTeX/KaTeX#696]
This may be related to #408. I have a fraction in a square root and there is no spacing at normal fontsize between the denominator and the vinculum of the fraction.

math in question: `\( K_m = K_e\sqrt{\frac{3}{2R}} \)`

KaTeX:
https://cdn.rawgit.com/jason-s/c913d7810ea9a72138eadbe3f87135dd/raw/1d4afa8441020a0245a463d12ec7d3f61e8c56b1/katex695.html (if you looked at this before May 11, please look again as I mistakenly posted a link to a MathJax implementation rather than KaTeX)
On my Lenovo W520 (1920x1080) running Windows 7, using Google Chrome:
![image](https://cloud.githubusercontent.com/assets/738893/25909277/8dc9545c-3561-11e7-8edc-8603b8850d2a.png)

On my Acer C720 Chromebook (1366 x 768) it does look ok at 100%:
![screenshot 2017-05-11 at 8 15 49 am](https://cloud.githubusercontent.com/assets/738893/25957003/213dc086-3622-11e7-9ef8-a390b42f9866.png)

but not at 70% where it seems to exhibit the same issue as on my Windows 7 laptop:

![screenshot 2017-05-11 at 8 24 33 am](https://cloud.githubusercontent.com/assets/738893/25957517/5d585742-3623-11e7-9186-60b5ba767ee9.png)

MathJax:
https://cdn.rawgit.com/jason-s/c913d7810ea9a72138eadbe3f87135dd/raw/37cf37aa54b4c83fe3c80e5fa013cab764173a3b/katex695-mathjax.html
![image](https://cloud.githubusercontent.com/assets/738893/25909710/e13cfb74-3562-11e7-9fd6-69b43c007936.png)

**Comment by jason-s** [comment]
At 150% zoom it looks ok:

![image](https://cloud.githubusercontent.com/assets/738893/25909979/be28173a-3563-11e7-9f55-485012d15751.png)

**Comment by jason-s** [comment]
(same example case as in #695)

**Comment by edemaine** [comment]
Can you add LaTeX rendering for comparison?

**Comment by jason-s** [comment]
hmm... will try, I have MiKTeX installed but I use it extremely rarely, just for TiKZ work.

**Comment by jason-s** [comment]
here we go:

```
\documentclass{article}
\begin{document}
$ K_m = K_e\sqrt{\frac{3}{2R}} $  (motor constant, Nm/root-W) 

$ \tau_e = L / R $ (electrical time constant, s) 

$ \tau_{m1} = J / K_m{}^2 $ (mechanical time constant, s)

$\xi = L_q / L_d $ (saliency ratio)
\end{document}
```

(courtesy of TikzEdt)
![image](https://cloud.githubusercontent.com/assets/738893/25911117/87293f62-3567-11e7-83fc-af0636174519.png)

**Comment by ronkok** [comment]
What follows is mostly speculation: In the relevant KaTeX source code, it’s clear that the objective was to match the kern distances described in the TeXbook. My guess is that the MathJax people looked at the result on a low-resolution monitor and decided to use a different spacing.

Here's how [QuickLaTeX](http://www.quicklatex.com/) renders it:

![Motor constant](http://quicklatex.com/cache3/db/ql_df7cb715e0ee0ad39791d7aa5ee52adb_l3.svg)
  
I think that KaTeX comes pretty close to LaTeX.  That's a good thing at high resolution.

**Comment by edemaine** [comment]
Hmm, yeah, the closer I look at the 150% zoom vs. the LaTeX rendering, the closer they look.  So I'm not sure there is a bug here (viewing LaTeX as the reference specification).  Really we should use the texcompare script to see.

(Unless there's a way to guarantee some kerns render as at least one pixel -- then I could see doing so to get things to look good at small zooms.  But I'm not sure that's possible.)

**Comment by jason-s** [comment]
>Hmm, yeah, the closer I look at the 150% zoom vs. the LaTeX rendering, the closer they look. So I'm not sure there is a bug here (viewing LaTeX as the reference specification). 

There is perhaps a philosophical issue here. TeX (and LaTeX), as far as I am aware, has always been aimed at high-end typography -- namely for printing or for viewing in applications that can zoom in and see equations in detail (dvi viewer, PDF, postscript). MathJax and KaTeX are specifically aimed at on-screen viewing. Following LaTeX output as a reference makes sense for large-size output, where quantization issues are not significant, but for small-size output, the readability of individual pixels is important. Furthermore, the default font size in browsers is closer to this small-size case rather than the large-size case. 

Not sure how to resolve this issue, given that KaTeX sticks to HTML+CSS, and pure scaling doesn't solve this problem.

**Comment by jason-s** [comment]
I wonder if this is a fonts issue; KaTeX looks pretty good on my Chromebook but has problems on my Windows 7 laptop. Both show up as KaTeX_Math rendered fonts... odd.

**Comment by edemaine** [comment]
@jason-s Luckily, with the increased prevalence of high-DPI screens, print and screen are converging (though there remains the emissive vs. reflective light).  But the advantage of having LaTeX as the reference implementation is that we actually can define what is (roughly) "correct" vs. "incorrect".  So if something works in LaTeX, I can expect it to work the same on KaTeX, and vice versa.  I for one wouldn't trade in this feature for a vague "usually looks nice on a screen".  (But small tweaks, like a minimum kern of 1px, could make sense.)

Are you running Chrome on both Chromebook and Windows 7?  (There are differences even between Chrome and Firefox.)  Are they identical DPI screens?  If so, it'd be interesting to compare screenshots.

**Comment by jason-s** [comment]
>Are you running Chrome on both Chromebook and Windows 7? (There are differences even between Chrome and Firefox.) 

Yes. (see updated issue description)

>Are they identical DPI screens? If so, it'd be interesting to compare screenshots.

hmm, not sure, also not sure if that's relevant; the [Lenovo W520 laptop](https://support.lenovo.com/us/en/solutions/pd015362) has a 15.6" 1920x1080 display, but I also have a 24" external monitor with 1920x1080 that I use and has the same issue. Same pixels, after all -- unless I'm missing something obvious.

Right now I am at work with an ASUS 27" 2560 x 1440 display and it has the same problem in Chrome at 100% resolution; here's a screenshot below, which again, I would expect to have the same pixel content independent of video cards or displays... the determining factor should be (?) the fonts that are loaded and how the browser (Chrome in this case) renders them.

![image](https://cloud.githubusercontent.com/assets/738893/25962930/66e631e2-3634-11e7-8cfa-31ba14735ac1.png)

**Comment by jason-s** [comment]
Just for kicks, here are a couple of different zoom factors in Chrome: (100% was in previous comment)

110%:
![image](https://cloud.githubusercontent.com/assets/738893/25963003/a89547e0-3634-11e7-9b20-f5bbf5baabfc.png)

125%:
![image](https://cloud.githubusercontent.com/assets/738893/25963018/ba41a89e-3634-11e7-890a-ea778658688a.png)

90%:
![image](https://cloud.githubusercontent.com/assets/738893/25963044/d78471fc-3634-11e7-9c1a-a1b81ba03dfd.png)

80%:
![image](https://cloud.githubusercontent.com/assets/738893/25963059/e5682304-3634-11e7-8fb9-c8b69f8c7f93.png)

75%:
![image](https://cloud.githubusercontent.com/assets/738893/25963074/f42c35b0-3634-11e7-81f8-c66680149676.png)

67%:
![image](https://cloud.githubusercontent.com/assets/738893/25963108/08902638-3635-11e7-9e94-9759b46281ed.png)

**Comment by jason-s** [comment]
Hmm, I tried experimenting in Chrome's developer tools console to tweak the CSS for the `.katex .mfrac .frac-line:before` rule:

was:
```
    border-bottom-style: solid;
    border-bottom-width: 1px;
    content: "";
    display: block;
```

tweaked to:
```
    border-top-style: solid;
    border-top-width: 0.05ex;
    padding-top: 0.07ex;
    content: "";
    display: block;
```

and removed the `.katex .mfrac .frac-line:after` rule which was the following (why is border-bottom-width in terms of `em` (width) rather than `ex` (height) ?)

```
    border-bottom-style: solid;
    border-bottom-width: .04em;
    content: "";
    display: block;
    margin-top: -1px;
```

and got something that sort of looks better but it gets quantized by Chrome to jostle up or down to join the numerator or denominator depending on the resolution.

I guess I should just blame Chrome's renderer (or the CSS spec) for the way it handles quantization and rounding to determine exactly where the pixels go.

**Comment by edemaine** [comment]
Yeah, it's messy.  Your CSS reminds me of a way to do the "minimum separate of one pixel", though.  Because [margins collapse](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Box_Model/Mastering_margin_collapsing), we can have one margin of 1px and another margin of the desired spacing, and collapse them to get the maximum.  [Here's a simple JSFiddle illustration](https://jsfiddle.net/edemaine/yap042se/).  Of course, we have no control over what happens when the whole page gets scaled down, but assuming &ge;100% scaling, we'd have at least a one-pixel gap, which only matters when using a low-resolution display.  Perhaps this is interesting?

**Comment by ylemkimon** [comment]
v0.9.0 was released recently, which includes multiple PR regarding `frac` lines. Could you check this is still an issue?

**Comment by jason-s** [comment]
https://cdn.rawgit.com/jason-s/c913d7810ea9a72138eadbe3f87135dd/raw/8cabc52f75e3b162eaff55d13eb8d4b34ccfce64/katex695-0.9.0.html

**Comment by jason-s** [comment]
Same on Acer C720 Chromebook:

125%:

![image](https://user-images.githubusercontent.com/738893/39202174-05a0dd2c-47a6-11e8-8592-bc01306f61fa.png)

110%:

![image](https://user-images.githubusercontent.com/738893/39202198-12303858-47a6-11e8-8df3-f5380cbe35e5.png)

100%:

![image](https://user-images.githubusercontent.com/738893/39202224-21735a7a-47a6-11e8-8e02-950cae75bb3a.png)

90%:

![image](https://user-images.githubusercontent.com/738893/39202246-2d45a0a6-47a6-11e8-9b16-0576169698a7.png)

80%:

![image](https://user-images.githubusercontent.com/738893/39202262-39643a6e-47a6-11e8-9050-c85cfbadf52c.png)

75%:

![image](https://user-images.githubusercontent.com/738893/39202280-44d868ca-47a6-11e8-918a-80ef912a60b4.png)

67%:

![image](https://user-images.githubusercontent.com/738893/39202291-4f74a08c-47a6-11e8-883e-ca820a714f59.png)

**Comment by kevinbarabash** [comment]
@jason-s can you post the TeX source so that it's easy to test the output on newer versions of KaTeX?
